mbed_build/_internal/config/assemble_build_config.py

```diff
@@ -0,0 +1,60 @@
+#
+# Copyright (C) 2020 Arm Mbed. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+"""Configuration assembly algorithm."""
+import itertools
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Iterable, Set
+
+from mbed_build._internal.config.config import Config
+from mbed_build._internal.config.cumulative_data import CumulativeData
+from mbed_build._internal.config.source import Source
+from mbed_build._internal.find_files import LabelFilter, filter_files, find_files
+
+
+@dataclass
+class BuildConfig:
+    """Representation of build configuration."""
+
+    config: Config
+    macros: Set[str]
+
+
+def assemble_config(mbed_target: str, mbed_program_directory: Path) -> BuildConfig:
+    """Assemble BuildConfig for given target and program directory.
+
+    The structure and configuration of MbedOS requires us to do multiple passes over
+    configuration files, as each pass might affect which configuration files should be included
+    in the final configuration.
+    """
+    target_source = Source.from_target(mbed_target, mbed_program_directory)
+    mbed_lib_files = find_files("mbed_lib.json", mbed_program_directory)
+    return _assemble_config(target_source, mbed_lib_files)
+
+
+def _assemble_config(target_source: Source, mbed_lib_files: Iterable[Path]) -> BuildConfig:
+    previous_cumulative_data = None
+    current_cumulative_data = CumulativeData.from_sources([target_source])
+    while True:
+        filtered_files = _filter_files(mbed_lib_files, current_cumulative_data)
+        mbed_lib_sources = [Source.from_mbed_lib(file, current_cumulative_data.labels) for file in filtered_files]
+        all_sources = list(itertools.chain([target_source], mbed_lib_sources))
```

Comment:
Are you using `chain` instead of `extend` here for performance reasons?

Reply:
Good point, it makes no sense at all. :)

Reply:
No, that's plain silly: https://github.com/ARMmbed/mbed-build/pull/29/commits/4fe2c5dd6db572fd119785a687f6d5811f8005ea